### PR TITLE
removes dependency on creating tasks for each replication operation

### PR DESCRIPTION
**Pull request by sp98**
We initially went with the tasks framework to run multiple tasks at the same time. Since ware not running multiple tasks, so removing dependency on tasks framework and calling the methods directly

Signed-off-by: Santosh Pillai <sapillai@redhat.com>

### CODE REVIEW

File: controllers/replication/replication.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package replication
+
+import (
+	"github.com/csi-addons/volume-replication-operator/pkg/client"
+	"google.golang.org/grpc/codes"
```
Comment (Madhu-1) [inline controllers/replication/replication.go:21]: new line to separate out the packages?

File: controllers/replication/replication.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+	resp, err := r.Params.Replication.ResyncVolume(
+		r.Params.VolumeID,
+		r.Params.Secrets,
+		r.Params.Parameters,
+	)
+
+	return &ReplicationResponse{Response: resp, Error: err}
+}
+
+func (r *Replication) HasKnownGRPCError(response *ReplicationResponse) bool {
```
Comment (Madhu-1) [inline controllers/replication/replication.go:97]: Make this as a method for ReplicationResponse and have error codes as input?

Comment (sp98) [reply controllers/replication/replication.go]: done.

File: controllers/replication/replication.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	"github.com/csi-addons/volume-replication-operator/pkg/client"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// Replication represents the instance of a single replication operation
+type Replication struct {
+	Params      CommonRequestParameters
+	KnownErrors []codes.Code
```
Comment (Madhu-1) [inline controllers/replication/replication.go:29]: this is not needed to be part of this anymore?

Comment (sp98) [reply controllers/replication/replication.go:29]: you mean `knownErrors` need not be part of `Replication` struct?  I think we can still keep it as looks like a legit field for Replication.  Or we can also use `KnownErrors` as variables in each methods like `markVolumeAsPrimary`

Comment (sp98) [reply controllers/replication/replication.go]: using `KnownErrors` as a variable.

File: controllers/volumereplication_controller.go [hunk 4]
```diff
@@ -296,42 +292,30 @@ func (r *VolumeReplicationReconciler) markVolumeAsPrimary(volumeReplicationObjec
...
-		},
+	replication := replication.Replication{
+		Params: c,
 	}
-	resp := tasks.RunAll(promoteVolumeTasks)
 
-	isKnownError := r.hasKnownGRPCError(logger, promoteVolumeTasks, resp)
+	resp := replication.Promote()
+
+	isKnownError := resp.HasKnownGRPCError(volumePromotionKnownErrors)
```
Comment (Madhu-1) [inline controllers/volumereplication_controller.go:301]: How about

```go
resp := replication.Promote()
if res.Error!=nil{
	isKnownError := resp.HasKnownGRPCError(volumePromotionKnownErrors)
	if !isKnownError {
			logger.Error(resp.Error, "failed to promote volume")
			setFailedPromotionCondition(&volumeReplicationObject.Status.Conditions, volumeReplicationObject.Generation)
			return resp.Error
	}
		// force promotion
	logger.Info("force promoting volume due to known grpc error", "error", resp.Error)
	replication.Force = true
	resp := replication.Promote()
	if resp.Error != nil {
		logger.Error(resp.Error, "failed to force promote volume")
		setFailedPromotionCondition(&volumeReplicationObject.Status.Conditions, volumeReplicationObject.Generation)
		return resp.Error
        }
}
	setPromotedCondition(&volumeReplicationObject.Status.Conditions, volumeReplicationObject.Generation)
	return nil
```

File: controllers/volumereplication_controller.go [hunk 5]
```diff
@@ -296,42 +292,30 @@ func (r *VolumeReplicationReconciler) markVolumeAsPrimary(volumeReplicationObjec
...
 
-	promoteVolumeTasks := []*tasks.TaskSpec{
-		{
-			Name: promoteVolume,
-			Task: replication.NewPromoteVolumeTask(c, false),
-			KnownErrors: []codes.Code{
-				codes.FailedPrecondition,
-			},
-		},
+	replication := replication.Replication{
```
Comment (Madhu-1) [inline controllers/volumereplication_controller.go:295]: lets not keep package name and variable name same.

File: controllers/volumereplication_controller.go [hunk 6]
```diff
@@ -417,30 +394,23 @@ func (r *VolumeReplicationReconciler) disableVolumeReplication(logger logr.Logge
...
-			KnownErrors: []codes.Code{
-				codes.NotFound,
-			},
-		},
+	replication := replication.Replication{
+		Params: c,
 	}
 
-	resp := tasks.RunAll(disableVolumeReplicationTasks)
+	resp := replication.Disable()
```
Comment (Madhu-1) [inline controllers/volumereplication_controller.go:401]:
```go
resp := replication.Disable()
if resp.Error!=nil{
	if isKnownError := resp.HasKnownGRPCError(disableReplicationKnownErrors); isKnownError {
		logger.Info("volume not found", "volumeID", volumeID)
		return nil
	}
	logger.Error(resp.Error, "failed to disable volume replication")
	return resp.Error
}
return nil
```

**Review by Madhu-1 — COMMENTED** [review]
Have small suggestions. LGTM

**Review by Madhu-1 — APPROVED** [review]

**Pull request merged**